### PR TITLE
SPUI-722 > Add referral stats

### PR DESCRIPTION
**Pull request by DenisMartych**

### CODE REVIEW

File: lib/app/shared/features/rewards/helper/reward_description_item.dart [hunk 1]
```diff
@@ -29,21 +28,9 @@ class RewardsDescriptionItem extends HookWidget {
       ),
       width: 287.w,
```
Comment (eli1stark) [inline lib/app/shared/features/rewards/helper/reward_description_item.dart:29]: fix

File: lib/app/shared/features/rewards/helper/set_reward_description_item.dart [hunk 2]
```diff
@@ -1,19 +1,74 @@
-import '../../../../../service/services/signal_r/model/campaign_response_model.dart';
+import 'package:flutter/material.dart';
+import 'package:simple_kit/simple_kit.dart';
 
+import '../../../../../service/services/signal_r/model/campaign_response_model.dart';
 import '../model/condition_type.dart';
+import 'set_reward_description_style.dart';
 
-String setRewardDescriptionItem(
+Widget setRewardDescriptionItem(
```
Comment (eli1stark) [inline lib/app/shared/features/rewards/helper/set_reward_description_item.dart:8]: rename set to create

File: lib/app/shared/features/rewards/view/rewards.dart [hunk 3]
```diff
@@ -1,61 +1,85 @@
...
 class Rewards extends HookWidget {
   const Rewards({Key? key}) : super(key: key);
 
   @override
   Widget build(BuildContext context) {
-    final campaign = useProvider(campaignNotipod(false));
     final colors = useProvider(sColorPod);
 
+    final rewardsData = useProvider(campaignAndReferralNotipod);
+    final rng = Random();
```
Comment (eli1stark) [inline lib/app/shared/features/rewards/view/rewards.dart:22]: use normal name

File: lib/app/shared/features/rewards/view/rewards.dart [hunk 4]
```diff
@@ -1,61 +1,85 @@
...
-                  rewardIndicatorComplete: setRewardIndicatorComplete(
-                    campaign[index].conditions!,
-                    colors,
+      child: Column(
+        children: [
+          Expanded(
+            child: ListView(
+              children: [
+                for (final item in rewardsData.campaigns) ...[
+                  if (item.conditions == null ||
```
Comment (eli1stark) [inline lib/app/shared/features/rewards/view/rewards.dart:34]: refactor to private function

File: lib/app/shared/features/rewards/view/rewards.dart [hunk 5]
```diff
@@ -1,61 +1,85 @@
...
+                      color: setBannerColor(
+                        rng.nextInt(3),
+                        colors,
+                      ),
+                      primaryText: item.title,
+                      secondaryText: item.description,
+                      imageUrl: item.imageUrl,
+                    ),
+                  ],
+                  if (item.conditions != null &&
```
Comment (eli1stark) [inline lib/app/shared/features/rewards/view/rewards.dart:47]: refactor to second private function that you can use in the first one

File: plugins/simple_kit/lib/src/referral_stats/simple_referral_stats_item.dart [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    this.baselineHeight = 40.0,
+    this.valueColor,
+    required this.value,
+    required this.text,
+  }) : super(key: key);
+
+  final String text;
+  final double baselineHeight;
+  final String value;
+  final Color? valueColor;
```
Comment (eli1stark) [inline plugins/simple_kit/lib/src/referral_stats/simple_referral_stats_item.dart:16]: sort

File: plugins/simple_kit/lib/src/referral_stats/simple_referral_stats_item.dart [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+          ),
+        ),
+        Baseline(
+          baseline: baselineHeight,
+          baselineType: TextBaseline.alphabetic,
+          child: Text(
+            value,
+            style: (valueColor != null)
+                ? sSubtitle3Style.copyWith(color: valueColor)
+                : sSubtitle3Style,
```
Comment (eli1stark) [inline plugins/simple_kit/lib/src/referral_stats/simple_referral_stats_item.dart:40]:
```dart
sSubtitle3Style.copyWith(
  color: valueColor ?? SColorsLight().black
)
```

**Review by eli1stark — CHANGES_REQUESTED** [review]

File: lib/app/shared/features/rewards/helper/reward_description_item.dart (hunk 1 above)
Comment (DenisMartych) [reply lib/app/shared/features/rewards/helper/reward_description_item.dart:29]: done

File: lib/app/shared/features/rewards/helper/set_reward_description_item.dart (hunk 2 above)
Comment (DenisMartych) [reply lib/app/shared/features/rewards/helper/set_reward_description_item.dart:8]: done

File: lib/app/shared/features/rewards/view/rewards.dart (hunk 3 above)
Comment (DenisMartych) [reply lib/app/shared/features/rewards/view/rewards.dart:22]: done

File: plugins/simple_kit/lib/src/referral_stats/simple_referral_stats_item.dart (hunk 6 above)
Comment (DenisMartych) [reply plugins/simple_kit/lib/src/referral_stats/simple_referral_stats_item.dart:16]: done

File: plugins/simple_kit/lib/src/referral_stats/simple_referral_stats_item.dart (hunk 7 above)
Comment (DenisMartych) [reply plugins/simple_kit/lib/src/referral_stats/simple_referral_stats_item.dart:40]: done

File: lib/app/shared/features/rewards/view/rewards.dart (hunk 4 above)
Comment (DenisMartych) [reply lib/app/shared/features/rewards/view/rewards.dart]: done

File: lib/app/shared/features/rewards/view/rewards.dart (hunk 5 above)
Comment (DenisMartych) [reply lib/app/shared/features/rewards/view/rewards.dart]: done

File: lib/app/shared/features/referral_stats/notifier/campaign_and_referral_notifier.dart [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import '../../../../../shared/logging/levels.dart';
+
+class CampaignAndReferralStatsNotifier
+    extends StateNotifier<CampaignAndReferralStatsModel> {
+  CampaignAndReferralStatsNotifier({
+    required this.read,
+    required this.campaignAndReferralStats,
+  }) : super(
+    const CampaignAndReferralStatsModel(
+      campaigns: <CampaignModel>[],
```
Comment (VovaMedvediev) [inline lib/app/shared/features/referral_stats/notifier/campaign_and_referral_notifier.dart:15]: I think it is necessarily to add type here

File: lib/app/shared/features/referral_stats/notifier/campaign_and_referral_notifier.dart [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+  final CampaignAndReferralStatsModel campaignAndReferralStats;
+  final Reader read;
+
+  static final _logger = Logger('ReferrerStatsNotifier');
+
+  Future<void> update(
+      CampaignAndReferralStatsModel campaignAndReferralStats,
+      ) async {
+    _logger.log(notifier, 'updateReferrerStats');
```
Comment (VovaMedvediev) [inline lib/app/shared/features/referral_stats/notifier/campaign_and_referral_notifier.dart:30]: function name is not equal to log function name

File: lib/app/shared/features/rewards/helper/create_reward_banner.dart [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+import 'package:flutter/material.dart';
+import 'package:simple_kit/simple_kit.dart';
+
+import '../../../../../service/services/signal_r/model/campaign_response_model.dart';
+import '../../../helpers/set_banner_colors.dart';
+import '../../market_details/helper/format_news_date.dart';
+import 'create_reward_detail.dart';
+import 'set_reward_indicator_complete.dart';
+
+Widget createRewardBanner(
```
Comment (VovaMedvediev) [inline lib/app/shared/features/rewards/helper/create_reward_banner.dart:12]: it is not ok to create functions which returns Widget. It is better to add Stateless Widget. https://stackoverflow.com/a/53234826

File: lib/app/shared/features/rewards/helper/create_reward_banner.dart [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    CampaignModel item,
+    Random rng,
+    SimpleColors colors,
+    ) {
+  if (item.conditions == null ||
+      (item.conditions != null &&
+          item.conditions!.isEmpty)) {
+    return SRewardBanner(
+      color: setBannerColor(
+        rng.nextInt(3),
```
Comment (VovaMedvediev) [inline lib/app/shared/features/rewards/helper/create_reward_banner.dart]: add informative const value. Currently I can't understand what is "3"

File: lib/app/shared/features/rewards/view/rewards.dart [hunk 12]
```diff
@@ -1,61 +1,49 @@
...
-                  ),
-                  imageUrl: campaign[index].imageUrl,
-                  circleAvatarColor: setBannerColor(index, colors),
-                  rewardDetail: createRewardDetail(
-                    campaign[index].conditions!,
-                  ),
-                  rewardIndicatorComplete: setRewardIndicatorComplete(
-                    campaign[index].conditions!,
-                    colors,
+      child: Column(
```
Comment (VovaMedvediev) [inline lib/app/shared/features/rewards/view/rewards.dart:21]: I think Column is redundant here. It has only 1 child

File: lib/app/shared/features/rewards/helper/create_reward_banner.dart (hunk 10 above)
Comment (eli1stark) [reply lib/app/shared/features/rewards/helper/create_reward_banner.dart:12]: agree

File: lib/app/shared/features/referral_stats/notifier/campaign_and_referral_notifier.dart (hunk 9 above)
Comment (DenisMartych) [reply lib/app/shared/features/referral_stats/notifier/campaign_and_referral_notifier.dart:30]: done

File: lib/app/shared/features/rewards/helper/create_reward_banner.dart (hunk 10 above)
Comment (DenisMartych) [reply lib/app/shared/features/rewards/helper/create_reward_banner.dart:12]: done

File: lib/app/shared/features/rewards/helper/create_reward_banner.dart (hunk 11 above)
Comment (DenisMartych) [reply lib/app/shared/features/rewards/helper/create_reward_banner.dart]: done

File: lib/app/shared/features/referral_stats/notifier/campaign_and_referral_notifier.dart (hunk 8 above)
Comment (DenisMartych) [reply lib/app/shared/features/referral_stats/notifier/campaign_and_referral_notifier.dart:15]: done

File: lib/app/shared/features/rewards/view/rewards.dart (hunk 12 above)
Comment (DenisMartych) [reply lib/app/shared/features/rewards/view/rewards.dart:21]: done

**Pull request merged**